src/lexer.c

```diff
@@ -0,0 +1,111 @@
+#include "lexer.h"
+
+#include <err.h>
+#include <ctype.h>
+#include <string.h>
+
+static void quoting_reset(struct quoting_state *q)
+{
+	memset(q, 0, sizeof(*q));
+}
+
+static bool inside_quoting(const struct quoting_state *q)
+{
+	return q->backslashed == true;
+}
+
+int lexer_init(struct lexer *l, FILE *input)
+{
+	if (charstream_init(&l->stream, input) < 0) {
+		warnx("%s: failed to init charstream", __FUNCTION__);
+		return -1;
+	}
+
+	quoting_reset(&l->quoting);
+
+	l->cur = NULL;
+
+	return 0;
+}
+
+static void discard_line(struct lexer *l)
+{
+	char c;
+
+	do {
+		c = charstream_read(&l->stream);
+	} while (c != CHARSTREAM_EOF && c != '\n');
+}
+
+static int lexer_consume_char(struct lexer *l, struct token *tok)
+{
+	char c = charstream_read(&l->stream);
+	bool is_quoted = inside_quoting(&l->quoting);
+
+	/* step_quoting */
+	if (l->quoting.backslashed > 0)
+		l->quoting.backslashed--;
```

Comment:
```suggestion
	if (l->quoting.backslashed)
		l->quoting.backslashed = false;
```
I'm not sure `true` always evaluate to >0

Reply:
My bad, I thought `backslashed` was an `int`.....